### PR TITLE
Register debug-cache-stats with a decorator

### PR DESCRIPTION
**Pull request by user202729**
While doing https://github.com/qutebrowser/qutebrowser/pull/4808#discussion_r288973940, I feel that it should be done properly. (although this is just a refactor)

------

This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4819)

### CODE REVIEW

**Comment by user202729** [comment]
Should I prevent two functions with identical name from being registered?

**Comment by user202729** [comment]
How can I satisfy mypy? What's the return value of `lru_cache` called?

**Comment by user202729** [comment]
How can I satisfy mypy? What's the return value of `lru_cache` called?

**Comment by The-Compiler** [comment]
The mypy issues in `configtypes.py` is due to an old PR I merged without type annotations - I fixed those in master in b2b4e437d4d0d3a52a3e46c3c640b6a4f8465935.

**Comment by user202729** [comment]
No, there are really problems with the code. Try doing a mypy check on this code: (with each value of the type `T`)

```
from typing import *
import functools
from functools import *

# T = Any  # ok, but too loose
# T = Callable  # no attribute "cache_info"
T = functools._lru_cache_wrapper  # ok, but undocumented

l: List[T] = []

@lru_cache()
def f(): pass

l.append(f)
l[0].cache_info()
```

The only reason why the current code can be type checked successfully is that I forget to type annotate the `debug_cache_stats` function so it isn't type-checked. If `-> None` is added then it raises an error.

---------

By the way, currently the `lru_cache` throws away the function parameter's type, only keep the return value type. Try doing a type check on this: (it's unexpected behavior that `a` and `b` doesn't raise an error)

```
from functools import *

@lru_cache()
def f(x: int) -> int:
    return x*2

a: int = f(1,2)
b: int = f('asdf')
c: str = f(1)
```

We could make the `register` function wrap the `lru_cache` and avoid having to decorate both `register` and `lru_cache` like this:

(TODO)

or

(TODO)

If that's cleaner I may implement that.

File: qutebrowser/misc/debugcachestats.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# The second element of each tuple should be a lru_cache wrapped function
+_CACHE_FUNCTIONS = []  # type: typing.List[typing.Tuple[str, typing.Any]]
+
+
+_T = TypeVar('_T', bound=Callable)
+
+
+def register(name: typing.Optional[str] = None) -> Callable[[_T], _T]:
+    """Register a lru_cache wrapped function to be used with debug_cache_stats.
+    """
```
Comment (user202729) [inline qutebrowser/misc/debugcachestats.py]: Flake8 complain at this line, although I have no idea how to fix it.

Comment (toofar) [reply qutebrowser/misc/debugcachestats.py]: """
    Register a lru_cache wrapped function to be used with debug_cache_stats.
    """
Probably. Putting it all on one line might work but I'm not sure the new line length limits apply to docstrings.

Comment (The-Compiler) [reply qutebrowser/misc/debugcachestats.py]: I'd just `s/to be used with/for/` and then the line is short enough to fit it all in one line :wink:

Comment (user202729) [reply qutebrowser/misc/debugcachestats.py]: But no other docstrings are formatted like that.

**Comment by The-Compiler** [comment]
This has a slight behavior change: The tab width cache is now always displayed for the last opened window, not the one `:debug-cache-stats` gets run in. However, I think that's an acceptable compromise.

Thanks!

**Comment by user202729** [comment]
No, each function object is only created once. Try running this:

```
#!/bin/python

import functools

class A:
    @functools.lru_cache()
    def a(self,i):
        pass

x=A()
y=A()
x.a(1)
y.a(2)
print(A.a.cache_info())
print(x.a.cache_info())
print(y.a.cache_info())
```

The 3 output values are the same.

**Pull request merged**